queries/sun_position.py

```diff
@@ -0,0 +1,623 @@
+    Examples:
+        '03 15' => datetime.time(3,15)
+        '34,9'  => 34.9
+        '     ' => None
+    """
+    if not pt_str.isspace():
+        if " " in pt_str:
+            # Column contains a time value, convert to datetime
+            hour, minute = pt_str.split()
+            return datetime.time(hour=int(hour) % 24, minute=int(minute) % 60)
+
+        if "," in pt_str:
+            # Column contains solar height at solar noon (in degrees)
+            return float(pt_str.replace(",", "."))
+
+    return None
+
+
+_SOLAR_ROW_TYPE = Dict[_SOLAR_POS_ENUM, _SOLAR_CELL_TYPE]
+_SOLAR_DICT_TYPE = Dict[
+    str, Dict[Union[str, datetime.date], Union[LatLonTuple, _SOLAR_ROW_TYPE]]
+]
+
+
+def _parse_almanak_hi_data(text: List[str]) -> _SOLAR_DICT_TYPE:
+    """
+    Parse text received from Almanak HÍ endpoint into usable dict.
+    """
+    data: _SOLAR_DICT_TYPE = {}
+    city: Optional[str] = None
+    month: Optional[int] = None
+
+    for line in text:
+        city_re = re.match(_ALMANAK_CITY_REGEX, line)
+        if city_re:
+            # Matched line containing name of city
+            city, lat, lon = city_re.groups()
+            city = capitalize_placename(city.lower())
+
+            # Initialize dict for Icelandic city (place)
+            data[city] = {"pos": _convert_dms_lat_lon_to_decimal(lat, lon)}
+
+        else:
+            sun_re = re.match(_ALMANAK_SUNPOS_REGEX, line)
+            if sun_re and city is not None:
+                # Matched line containing times of solar positions
+                if sun_re.group(1) != " - ":
+                    # New month started
+                    month_str = sun_re.group(1).lower()
+                    month = MONTH_ABBREV.index(month_str) + 1
+
+                # Extract times of solar positions
+                sun_pos: _SOLAR_ROW_TYPE = dict(
+                    zip(
+                        _ALMANAK_HI_COLUMNS,
+                        (_parse_almanak_cell(cell) for cell in sun_re.groups()[2:]),
+                    )
+                )
+
+                # Calculate solar midnight from solar noon
+                # ("Hádegi" (solar noon) is never None)
+                solar_noon: datetime.time = cast(
+                    datetime.time, sun_pos[_SOLAR_POSITIONS.HÁDEGI]
+                )
+                solar_midnight = datetime.time(
+                    hour=((solar_noon.hour + 12) % 24), minute=solar_noon.minute
+                )
+                sun_pos[_SOLAR_POSITIONS.MIÐNÆTTI] = solar_midnight
+
+                # Add solar positions for city on a specific date
+                date = datetime.date.today().replace(
+                    month=cast(int, month), day=int(sun_re.group(2))
+                )
+                data[city][date] = sun_pos
+
+    return data
+
+
+@cached(TTLCache(maxsize=1, ttl=86400))
+def _get_almanak_hi_data() -> _SOLAR_DICT_TYPE:
+    r = requests.get(_ALMANAK_HI_URL)
```

Comment:
Be careful with potential exceptions here. If there is a problem fetching the data, we want an orderly recovery and retry, at least within some elapsed time.

Reply:
Added a few try-except blocks similar to the ones in query_json_api.